service/history/decisionHandler.go

```diff
@@ -456,11 +456,29 @@ Update_History_Loop:
 			continueAsNewBuilder = nil
 		}
 
-		// Schedule another decision task if new events came in during this decision or if request forced to
-		createNewDecisionTask := msBuilder.IsWorkflowExecutionRunning() &&
-			(hasUnhandledEvents || request.GetForceCreateNewDecisionTask() || activityNotStartedCancelled)
+		queryResults := req.GetCompleteRequest().GetQueryResults()
+		bufferedQueries := msBuilder.GetQueryRegistry().getBufferedSnapshot()
+		hasUnhandledQueries := false
+		if len(queryResults) < len(bufferedQueries) {
+			hasUnhandledQueries = true
+		} else {
+			for _, bid := range bufferedQueries {
+				if _, ok := queryResults[bid]; !ok {
+					hasUnhandledQueries = true
+					break
+				}
+			}
+		}
+
+		createNewDecisionTask := msBuilder.IsWorkflowExecutionRunning() && (hasUnhandledEvents || request.GetForceCreateNewDecisionTask() || activityNotStartedCancelled || hasUnhandledQueries)
```

Comment:
this maybe a long term thing, but for respond decision task completed, plz keep in mind that the code is still hard to read, and we need to further refactor this.